[Bug] Snapshot file is erroneously deleted during FS rename retry, causing FileNotFoundException upon Flink failover

### Search before asking

- [x] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.


### Paimon version

0.8.2

### Compute Engine

Flink

### Minimal reproduce step

1. Run a Flink job writing to Paimon on Object Storage (S3).

2. Simulate a network timeout specifically during the "Commit/Rename" phase of a checkpoint.

3. Trigger the file system client to retry the rename operation.

4. Observe the deletion of the target snapshot file.

5. Wait for Flink failover; the job will get stuck in a restart loop with FileNotFoundException.
`2025-11-12 18:05:09
java.lang.RuntimeException: java.io.FileNotFoundException: File 's3://dt-warehouse/paimon-warehouse/nd_game_sjmy_cdm.db/dwd_sjmy_gmlog_boss_user_bekilllist/manifest/manifest-list-20f67c4d-ba05-4dbf-a26a-24f229b949cc-32' not found, Possible causes: 1.snapshot expires too fast, you can configure 'snapshot.time-retained' option with a larger value. 2. consumption is too slow, you can improve the performance of consumption (For example, increasing parallelism).
	at org.apache.paimon.flink.sink.AsyncLookupSinkWrite.<init>(AsyncLookupSinkWrite.java:75)
	at org.apache.paimon.flink.sink.FlinkSink.lambda$createWriteProvider$672a9d60$1(FlinkSink.java:147)
	at org.apache.paimon.flink.sink.TableWriteOperator.initializeState(TableWriteOperator.java:78)
	at org.apache.flink.streaming.api.operators.StreamOperatorStateHandler.initializeOperatorState(StreamOperatorStateHandler.java:122)
	at org.apache.flink.streaming.api.operators.AbstractStreamOperator.initializeState(AbstractStreamOperator.java:316)
	at org.apache.flink.streaming.api.operators.AbstractStreamOperator.initializeState(AbstractStreamOperator.java:306)
	at org.apache.flink.streaming.runtime.tasks.RegularOperatorChain.initializeStateAndOpenOperators(RegularOperatorChain.java:107)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.restoreGates(StreamTask.java:759)
	at org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor$1.call(StreamTaskActionExecutor.java:55)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.restoreInternal(StreamTask.java:734)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.restore(StreamTask.java:699)
	at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:971)
	at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:940)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:764)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:574)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.RuntimeException: java.io.FileNotFoundException: File 's3://dt-warehouse/paimon-warehouse/nd_game_sjmy_cdm.db/dwd_sjmy_gmlog_boss_user_bekilllist/manifest/manifest-list-20f67c4d-ba05-4dbf-a26a-24f229b949cc-32' not found, Possible causes: 1.snapshot expires too fast, you can configure 'snapshot.time-retained' option with a larger value. 2. consumption is too slow, you can improve the performance of consumption (For example, increasing parallelism).
	at org.apache.paimon.utils.ObjectsCache.readSegments(ObjectsCache.java:143)
	at org.apache.paimon.utils.ObjectsCache.read(ObjectsCache.java:93)
	at org.apache.paimon.utils.ObjectsFile.readWithIOException(ObjectsFile.java:149)
	at org.apache.paimon.utils.ObjectsFile.read(ObjectsFile.java:134)
	at org.apache.paimon.utils.ObjectsFile.read(ObjectsFile.java:105)
	at org.apache.paimon.manifest.ManifestList.readDataManifests(ManifestList.java:90)
	at org.apache.paimon.operation.ManifestsReader.readManifests(ManifestsReader.java:128)
	at org.apache.paimon.operation.ManifestsReader.read(ManifestsReader.java:114)
	at org.apache.paimon.operation.AbstractFileStoreScan.readManifests(AbstractFileStoreScan.java:417)
	at org.apache.paimon.operation.AbstractFileStoreScan.plan(AbstractFileStoreScan.java:257)
	at org.apache.paimon.operation.AbstractFileStoreWrite.scanExistingFileMetas(AbstractFileStoreWrite.java:491)
	at org.apache.paimon.operation.AbstractFileStoreWrite.createWriterContainer(AbstractFileStoreWrite.java:440)`

### What doesn't meet your expectations?

The rename operation should be idempotent or check for the existence of the target file before considering the operation failed. If the source is missing but the target exists and matches expectations during a retry, it should be treated as a success, or at least the target file should not be deleted.

### Anything else?

_No response_

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/paimon-core/src/main/java/org/apache/paimon/catalog/RenamingSnapshotCommit.java
+++ b/paimon-core/src/main/java/org/apache/paimon/catalog/RenamingSnapshotCommit.java
@@ -25,6 +25,7 @@ import org.apache.paimon.operation.Lock;
 import org.apache.paimon.partition.PartitionStatistics;
 import org.apache.paimon.utils.SnapshotManager;
 
+import java.io.IOException;
 import java.util.List;
 import java.util.concurrent.Callable;
 
@@ -57,6 +58,20 @@ public class RenamingSnapshotCommit implements SnapshotCommit {
         Callable<Boolean> callable =
                 () -> {
                     boolean committed = fileIO.tryToWriteAtomic(newSnapshotPath, snapshot.toJson());
+                    if (!committed) {
+                        if (!fileIO.exists(newSnapshotPath)) {
+                            throw new IOException(
+                                    "Commit snapshot "
+                                            + snapshot.id()
+                                            + " failed and "
+                                            + newSnapshotPath
+                                            + " not found");
+                        }
+                        committed =
+                                snapshot.equals(
+                                        Snapshot.fromJson(fileIO.readFileUtf8(newSnapshotPath)));
+                    }
+
                     if (committed) {
                         snapshotManager.commitLatestHint(snapshot.id());
                     }

--- a/paimon-core/src/test/java/org/apache/paimon/catalog/RenamingSnapshotCommitTest.java
+++ b/paimon-core/src/test/java/org/apache/paimon/catalog/RenamingSnapshotCommitTest.java
@@ -1,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.catalog;
+
+import org.apache.paimon.Snapshot;
+import org.apache.paimon.fs.FileIO;
+import org.apache.paimon.fs.Path;
+import org.apache.paimon.fs.local.LocalFileIO;
+import org.apache.paimon.operation.Lock;
+import org.apache.paimon.utils.HintFileUtils;
+import org.apache.paimon.utils.SnapshotManager;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/**
+ * Tests for {@link RenamingSnapshotCommit} to verify robustness when object storage rename returns
+ * false while the target file is actually created, and to verify failure when target is missing.
+ */
+public class RenamingSnapshotCommitTest {
+
+    /**
+     * Simulate the object storage scenario where rename actually succeeds but returns false, which
+     * causes {@link FileIO#tryToWriteAtomic(Path, String)} to return false while the target file
+     * exists with correct content.
+     */
+    @Test
+    public void testCommitSucceedsAndRenameReturnsFalse(@TempDir java.nio.file.Path tmp)
+            throws Exception {
+        FileIO fileIO = new AlwaysFalseRenameLocalFileIO(true);
+        Path tablePath = new Path(tmp.toUri());
+        SnapshotManager snapshotManager = new SnapshotManager(fileIO, tablePath, null, null, null);
+
+        RenamingSnapshotCommit commit = new RenamingSnapshotCommit(snapshotManager, Lock.empty());
+
+        Snapshot snapshot = createSnapshot(1L);
+
+        boolean committed = commit.commit(snapshot, "main", Collections.emptyList());
+
+        assertThat(committed).isTrue();
+
+        // snapshot file should exist and match content
+        Path snapshotPath = snapshotManager.snapshotPath(snapshot.id());
+        assertThat(fileIO.exists(snapshotPath)).isTrue();
+        Snapshot loaded = Snapshot.fromJson(fileIO.readFileUtf8(snapshotPath));
+        assertThat(loaded).isEqualTo(snapshot);
+
+        // LATEST hint should be committed
+        Path latestHint = new Path(snapshotManager.snapshotDirectory(), HintFileUtils.LATEST);
+        assertThat(fileIO.readOverwrittenFileUtf8(latestHint).orElse(null))
+                .isEqualTo(String.valueOf(snapshot.id()));
+    }
+
+    /**
+     * Simulate the scenario where rename fails and the target snapshot file is not created; the
+     * commit should throw IOException according to the new check.
+     */
+    @Test
+    public void testCommitTargetSnapshotMissing(@TempDir java.nio.file.Path tmp) throws Exception {
+        FileIO fileIO = new AlwaysFalseRenameLocalFileIO(false);
+        Path tablePath = new Path(tmp.toUri());
+        SnapshotManager snapshotManager = new SnapshotManager(fileIO, tablePath, null, null, null);
+
+        RenamingSnapshotCommit commit = new RenamingSnapshotCommit(snapshotManager, Lock.empty());
+
+        Snapshot snapshot = createSnapshot(2L);
+
+        IOException ex =
+                assertThrows(
+                        IOException.class,
+                        () -> commit.commit(snapshot, "main", Collections.emptyList()));
+
+        assertThat(ex)
+                .hasMessageContaining("Commit snapshot " + snapshot.id() + " failed")
+                .hasMessageContaining("not found");
+
+        // ensure snapshot file is not accidentally created
+        assertThat(fileIO.exists(snapshotManager.snapshotPath(snapshot.id()))).isFalse();
+    }
+
+    private static Snapshot createSnapshot(long id) throws IOException {
+        long schemaId = 1L;
+        String baseManifestList = "manifest-list-base";
+        String deltaManifestList = "manifest-list-delta";
+        String changelogManifestList = null;
+        Long baseManifestListSize = 10L;
+        Long deltaManifestListSize = 20L;
+        Long changelogManifestListSize = null;
+        String indexManifest = null;
+        String commitUser = "user";
+        long commitIdentifier = id;
+        Snapshot.CommitKind commitKind = Snapshot.CommitKind.APPEND;
+        long timeMillis = System.currentTimeMillis();
+        Map<Integer, Long> logOffsets = new HashMap<>();
+        Long totalRecordCount = 100L;
+        Long deltaRecordCount = 100L;
+
+        return new Snapshot(
+                id,
+                schemaId,
+                baseManifestList,
+                baseManifestListSize,
+                deltaManifestList,
+                deltaManifestListSize,
+                changelogManifestList,
+                changelogManifestListSize,
+                indexManifest,
+                commitUser,
+                commitIdentifier,
+                commitKind,
+                timeMillis,
+                logOffsets,
+                totalRecordCount,
+                deltaRecordCount,
+                null,
+                null,
+                null,
+                null,
+                null);
+    }
+
+    /**
+     * A FileIO based on LocalFileIO that always returns false from rename().
+     *
+     * <p>When {@code actuallyMove} is true, it performs the move then returns false; otherwise it
+     * returns false without moving, simulating a failed rename.
+     */
+    private static class AlwaysFalseRenameLocalFileIO extends LocalFileIO {
+
+        private final boolean actuallyMove;
+
+        AlwaysFalseRenameLocalFileIO(boolean actuallyMove) {
+            this.actuallyMove = actuallyMove;
+        }
+
+        @Override
+        public boolean rename(Path src, Path dst) throws IOException {
+            if (actuallyMove) {
+                // perform actual move to create target, but report failure
+                boolean ignored = super.rename(src, dst);
+                return false;
+            } else {
+                // do not move, report failure
+                // ensure parent directory exists to avoid side effects in tests
+                java.nio.file.Path parent = java.nio.file.Paths.get(dst.toUri());
+                Files.createDirectories(parent.getParent());
+                return false;
+            }
+        }
+    }
+}

